pkg/scan/docker/docker.go

```diff
@@ -0,0 +1,152 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/mholt/archiver/v3"
+	"github.com/palantir/log4j-sniffer/pkg/archive"
+	"github.com/palantir/log4j-sniffer/pkg/crawl"
+	"github.com/palantir/log4j-sniffer/pkg/scan"
+	"github.com/pkg/errors"
+)
+
+type Scanner struct {
+	config     scan.Config
+	crawler    crawl.Crawler
+	reporter   *crawl.Reporter
+	identifier crawl.Identifier
+	client     client.CommonAPIClient
+}
+
+func NewDockerScanner(config scan.Config, stdout, stderr io.Writer) (*Scanner, error) {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create docker client")
+	}
+
+	return &Scanner{
+		config: config,
+		crawler: crawl.Crawler{
+			ErrorWriter: stderr,
+			IgnoreDirs:  config.Ignores,
+		},
+		reporter: &crawl.Reporter{
+			OutputJSON:      config.OutputJSON,
+			OutputWriter:    stdout,
+			DisableCVE45105: config.DisableCVE45105,
+		},
+		identifier: crawl.NewIdentifier(config.ArchiveListTimeout, archive.WalkZipFiles, archive.WalkTarGzFiles),
+		client:     dockerClient,
+	}, nil
+}
+
+func (d Scanner) ScanImages(ctx context.Context) (int64, error) {
+	imageList, err := d.client.ImageList(ctx, dockertypes.ImageListOptions{})
+	if err != nil {
+		return 0, errors.Wrap(err, "could not list docker images")
+	}
+
+	var stats crawl.Stats
+	for _, image := range imageList {
+		if len(image.RepoTags) == 0 {
+			continue
+		}
+		d.reporter.SetImageID(image.ID)
+		d.reporter.SetImageTags(image.RepoTags)
+		imageStats, err := d.scanImage(ctx, image)
+		if err != nil {
+			// write an error and continue scanning other images if we hit an error with this
+			// image
+			_, _ = fmt.Fprintln(d.crawler.ErrorWriter, err.Error())
+			continue
+		}
+		stats.Append(imageStats)
+	}
+
+	count := d.reporter.Count()
+	if d.config.OutputSummary {
+		if err := scan.WriteSummary(d.reporter.OutputWriter, d.config, stats, count); err != nil {
+			return 0, err
+		}
+	}
+	return count, nil
+}
+
+func (d Scanner) scanImage(ctx context.Context, image dockertypes.ImageSummary) (crawl.Stats, error) {
+	ref, err := name.ParseReference(image.RepoTags[0])
+	if err != nil {
+		return crawl.Stats{}, errors.Wrapf(err, "failed to get image reference")
+	}
+
+	img, err := daemon.Image(ref, daemon.WithClient(d.client), daemon.WithContext(ctx))
+	if err != nil {
+		return crawl.Stats{}, err
+	}
+
+	// create a temporary directory where the docker image tarball can be exported to
+	imageTmpDir, err := os.MkdirTemp("", fmt.Sprintf("log4j-sniffer-%s", image.ID))
+	if err != nil {
+		return crawl.Stats{}, errors.Wrap(err, "could not create temporary directory for image")
+	}
```

Comment:
Do we want to give control to the user to configure where to write any temporary files we will create?
This will be useful for some of the unarchiving work where we want to ensure we are keeping a cap on the memory profile for large archives.

Reply:
This makes sense, have added a `--image-dir` flag to the docker command to allow this to be configured